tower-http-tls/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+extern crate futures;
+extern crate http;
+extern crate rustls;
+extern crate tokio_io;
+extern crate tokio_rustls;
+extern crate tokio_tcp;
+extern crate tower_http_util;
+extern crate tower_service;
+extern crate webpki;
+extern crate webpki_roots;
+
+use futures::{Future, Poll};
+use http::{
+    uri::{Scheme, Uri},
+    Version,
+};
+use rustls::{ClientConfig, Session};
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::sync::Arc;
+use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_rustls::client::TlsStream;
+use tokio_tcp::TcpStream;
+use tower_http_util::HttpConnection;
+use tower_service::Service;
+use webpki::DNSNameRef;
+use webpki_roots::TLS_SERVER_ROOTS;
+
+const ALPN_H2: &str = "h2";
+
+pub struct TlsConnector {
+    config: Arc<ClientConfig>,
+}
+
+pub struct TlsConnection {
+    inner: TlsStream<TcpStream>,
+}
+
+impl TlsConnector {
+    pub fn new(config: ClientConfig) -> Self {
+        let config = Arc::new(config);
+        TlsConnector { config }
+    }
+
+    pub fn with_root(h2: bool) -> Self {
```

Comment:
Should probably allow an option to specify your own root cert, and also a way to pass a client cert/key pair.

Reply:
The `new` takes a ClientConfig so you can do it that way. I believe this just pulls from the default Mozilla root cert. Though we could prob add more until constructors around that.

Reply:
Fair enough. I thought I'd mention it anyway since I assume this is for use with gRPC, and MTLS is quite common with gRPC nowadays (for good reason).